src/posix/platform/spi_interface.hpp

```diff
@@ -0,0 +1,228 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for the SPI interface to radio (RCP).
+ */
+
+#ifndef POSIX_APP_SPI_INTERFACE_HPP_
+#define POSIX_APP_SPI_INTERFACE_HPP_
+
+#include "spinel_interface.hpp"
+#include "ncp/hdlc.hpp"
+
+#include <openthread-system.h>
+
+#if OPENTHREAD_POSIX_RCP_SPI_ENABLE
+
+#include "ncp/ncp_spi.hpp"
+
+namespace ot {
+namespace PosixApp {
+
+/**
+ * This class defines an SPI interface to the Radio Co-processor (RCP)
+ *
+ */
+class SpiInterface
+{
+public:
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in] aCallback     A reference to a `Callback` object.
+     * @param[in] aFrameBuffer  A reference to a `RxFrameBuffer` object.
+     *
+     */
+    SpiInterface(SpinelInterface::Callbacks &aCallback, SpinelInterface::RxFrameBuffer &aFrameBuffer);
+
+    /**
+     * This destructor deinitializes the object.
+     *
+     */
+    ~SpiInterface(void);
+
+    /**
+     * This method initializes the interface to the Radio Co-processor (RCP)
+     *
+     * @note This method should be called before reading and sending spinel frames to the interface.
+     *
+     * @param[in]  aPlatformConfig  Platform configuration structure.
+     *
+     * @retval OT_ERROR_NONE          The interface is initialized successfully
+     * @retval OT_ERROR_ALREADY       The interface is already initialized.
+     * @retval OT_ERROR_INVALID_ARGS  The UART device or executable cannot be found or failed to open/run.
+     *
+     */
+    otError Init(const otPlatformConfig &aPlatformConfig);
+
+    /**
+     * This method deinitializes the interface to the RCP.
+     *
+     */
+    void Deinit(void);
+
+    /**
+     * This method encodes and sends a spinel frame to Radio Co-processor (RCP) over the socket.
+     *
+     * @param[in] aFrame     A pointer to buffer containing the spinel frame to send.
+     * @param[in] aLength    The length (number of bytes) in the frame.
+     *
+     * @retval OT_ERROR_NONE     Successfully encoded and sent the spinel frame.
+     * @retval OT_ERROR_BUSY     Failed due to another operation is on going.
+     * @retval OT_ERROR_NO_BUFS  Insufficient buffer space available to encode the frame.
+     * @retval OT_ERROR_FAILED   Failed to call the SPI driver to send the frame.
+     *
+     */
+    otError SendFrame(const uint8_t *aFrame, uint16_t aLength);
+
+    /**
+     * This method waits for receiving part or all of spinel frame within specified interval.
+     *
+     * @param[inout]  aTimeout  A reference to the timeout.
+     *
+     * @retval OT_ERROR_NONE             Part or all of spinel frame is received.
+     * @retval OT_ERROR_RESPONSE_TIMEOUT No spinel frame is received within @p aTimeout.
```

Comment:
I think there will be an interrupted situation.

Reply:
The posix host doesn't use `signal`. So we don't need to consider the interrupt.